File: codes/matlab/array_find_index.m
function stringifyresult = stringify(parameter)
    if isempty(parameter)
        stringifyresult = "null";
        return;
    end
    if isstring(parameter)
        stringifyresult = strcat("""", parameter, """");
        return;
    end
    if ischar(parameter)
        stringifyresult = strcat("""", parameter, """");
        return;
    end
    if isnumeric(parameter)
        stringifyresult = num2str(parameter);
        return;
    end
    if (islogical(parameter) && parameter == true)
        stringifyresult = "true";
        return;
    end
    if (islogical(parameter) && parameter == false)
        stringifyresult = "false";
        return;
    end
    if iscell(parameter)
        if isempty(parameter)
            stringifyresult = "[]";
            return;
        end
        currentresult = "[";
        for arrayitemindex = (1:1:numel(parameter))
            arrayitem = parameter{arrayitemindex};
            currentresult = strcat(currentresult, stringify(arrayitem));
            if (arrayitemindex ~= numel(parameter))
                currentresult = strcat(currentresult, ", ");
            end
        end
        stringifyresult = strcat(currentresult, "]");
        return;
    end
    if isstruct(parameter)
        if (numel(fieldnames(parameter)) == 0)
            stringifyresult = "{}";
            return;
        end
        currentresult = "{";
        objectkeys = fieldnames(parameter);
        for entryindex = (1:1:numel(objectkeys))
            objectkey = objectkeys{entryindex};
            objectvalue = parameter.(objectkey);
            currentresult = strcat(currentresult, """", objectkey, """: ", stringify(objectvalue));
            if (entryindex ~= numel(objectkeys))
                currentresult = strcat(currentresult, ",");
            end
        end
        stringifyresult = strcat(currentresult, "}");
        return;
    end
    stringifyresult = "null";
end

function sprint(varargin)
    currentresult = "";
    for parameterindex = (1:1:numel(varargin))
        parameter = varargin{parameterindex};
        if (iscell(parameter) && (numel(parameter) == 1))
            parameternew = stringify(parameter{1});
            parameternew = strrep(parameternew, """{", "{");
            parameternew = strrep(parameternew, """[", "[");
            parameternew = strrep(parameternew, "}""", "}");
            parameternew = strrep(parameternew, "]""", "]");
            currentresult = strcat(currentresult, parameternew);
            continue;
        end
        currentresult = strcat(currentresult, parameter);
    end
    disp(currentresult);
end

function srepeatresult = srepeat(astring, count)
    result = "";
    for i = (1:1:count)
        result = strcat(result, astring);
    end
    srepeatresult = result;
end

function prettyjsonstringifyresult = prettyjsonstringify(parameter)
    if ((nargin == 0) || (~iscell(parameter)) || (iscell(parameter) && isempty(parameter)))
        error("Argument should be wrapped in {}");
    end
    indent = "    ";
    if ((nargin == 2) && isstruct(parameter{2}) && isfield(parameter{2}, "indent"))
        indent = parameter{2}.indent;
    end
    anything = parameter{1};
    indentlevel = 0;
    function prettyjsonstringifyinnerresult = prettyjsonstringifyinner(anythinginner, indentinner)
        if isempty(anythinginner)
            prettyjsonstringifyinnerresult = "null";
            return;
        end
        if isstring(anythinginner)
            prettyjsonstringifyinnerresult = strcat("""", anythinginner, """");
            return;
        end
        if ischar(anythinginner)
            prettyjsonstringifyinnerresult = strcat("""", anythinginner, """");
            return;
        end
        if isnumeric(anythinginner)
            prettyjsonstringifyinnerresult = num2str(anythinginner);
            return;
        end
        if (islogical(anythinginner) && anythinginner == true)
            prettyjsonstringifyinnerresult = "true";
            return;
        end
        if (islogical(anythinginner) && anythinginner == false)
            prettyjsonstringifyinnerresult = "false";
            return;
        end
        if iscell(anythinginner)
            if isempty(anythinginner)
                prettyjsonstringifyinnerresult = "[]";
                return;
            end
            indentlevel = indentlevel + 1;
            result = strcat("[", sprintf("\n"), srepeat(indentinner, indentlevel));
            for arrayitemindex = (1:1:numel(anythinginner))
                arrayitem = anythinginner{arrayitemindex};
                result = strcat(result, prettyjsonstringifyinner(arrayitem, indentinner));
                if (arrayitemindex ~= numel(anythinginner))
                    result = strcat(result, ",", sprintf("\n"), srepeat(indentinner, indentlevel));
                end
            end
            indentlevel = indentlevel - 1;
            result = strcat(result, sprintf("\n"), srepeat(indentinner, indentlevel), "]");
            prettyjsonstringifyinnerresult = result;
            return;
        end
        if isstruct(anythinginner)
            if (numel(fieldnames(anythinginner)) == 0)
                prettyjsonstringifyinnerresult = "{}";
                return;
            end
            indentlevel = indentlevel + 1;
            result = strcat("{", sprintf("\n"), srepeat(indentinner, indentlevel));
            objectkeys = fieldnames(anythinginner);
            for entryindex = (1:1:numel(objectkeys))
                objectkey = objectkeys{entryindex};
                objectvalue = anythinginner.(objectkey);
                result = strcat(result, """", objectkey, """: ", prettyjsonstringifyinner(objectvalue, indentinner));
                if (entryindex ~= numel(objectkeys))
                    result = strcat(result, ",", sprintf("\n"), srepeat(indentinner, indentlevel));
                end
            end
            indentlevel = indentlevel - 1;
            result = strcat(result, sprintf("\n"), srepeat(indentinner, indentlevel), "}");
            prettyjsonstringifyinnerresult = result;
            return;
        end
        prettyjsonstringifyinnerresult = "null";
    end
    prettyjsonstringifyresult = prettyjsonstringifyinner(anything, indent);
end

function arrayfindindexv1result = arrayfindindexv1(callbackfunction, anarray)
    % JavaScript-like Array.findIndex() function
    datafoundindex = -1;
    for arrayitemindex = (1:1:numel(anarray))
        arrayitem = anarray{arrayitemindex};
        isconditionmatch = callbackfunction(arrayitem, arrayitemindex, anarray);
        if (isconditionmatch == true)
            datafoundindex = arrayitemindex;
            break;
        end
    end
    arrayfindindexv1result = datafoundindex;
end

function arrayfindindexv2result = arrayfindindexv2(callbackfunction, anarray)
    % JavaScript-like Array.findIndex() function
    datafoundindex = -1;
    for arrayitemindex = (1:1:numel(anarray))
        arrayitem = anarray{arrayitemindex};
        if (callbackfunction(arrayitem, arrayitemindex, anarray) == true)
            datafoundindex = arrayitemindex;
            break;
        end
    end
    arrayfindindexv2result = datafoundindex;
end

function arrayfindindexv3result = arrayfindindexv3(callbackfunction, anarray)
    % JavaScript-like Array.findIndex() function
    for arrayitemindex = (1:1:numel(anarray))
        arrayitem = anarray{arrayitemindex};
        isconditionmatch = callbackfunction(arrayitem, arrayitemindex, anarray);
        if (isconditionmatch == true)
            arrayfindindexv3result = arrayitemindex;
            return;
        end
    end
    arrayfindindexv3result = -1;
end

function arrayfindindexv4result = arrayfindindexv4(callbackfunction, anarray)
    % JavaScript-like Array.findIndex() function
    for arrayitemindex = (1:1:numel(anarray))
        arrayitem = anarray{arrayitemindex};
        if (callbackfunction(arrayitem, arrayitemindex, anarray) == true)
            arrayfindindexv4result = arrayitemindex;
            return;
        end
    end
    arrayfindindexv4result = -1;
end

sprint(sprintf("\n"), "% JavaScript-like Array.findIndex() in Matlab Cell Array");

numbers = {12, 34, 27, 23, 65, 93, 36, 87, 4, 254};
sprint("numbers: ", {numbers});

numbertofind = 27;
sprint("number to find: ", {numbertofind});

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv1""");

numberfoundindex = arrayfindindexv1(@(number, varargin) (number == numbertofind), numbers);
sprint("number found index: ", {numberfoundindex});
% number found index: 3

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv2""");

numberfoundindex = arrayfindindexv2(@(number, varargin) (number == numbertofind), numbers);
sprint("number found index: ", {numberfoundindex});
% number found index: 3

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv3""");

numberfoundindex = arrayfindindexv3(@(number, varargin) (number == numbertofind), numbers);
sprint("number found index: ", {numberfoundindex});
% number found index: 3

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv4""");

numberfoundindex = arrayfindindexv4(@(number, varargin) (number == numbertofind), numbers);
sprint("number found index: ", {numberfoundindex});
% number found index: 3

sprint("% using Matlab Array.findIndex() built-in function ""find"" and ""cellfun""");

numberfoundindex = find(cellfun(@(number) (number == numbertofind), numbers), 1);
sprint("number found index: ", {numberfoundindex});
% number found index: 3

sprint(sprintf("\n"), "% JavaScript-like Array.findIndex() in Matlab Cell Array of Structs");

products = { ...
    struct( ...
        "code", "pasta", ...
        "price", 321 ...
    ), ...
    struct( ...
        "code", "bubble_gum", ...
        "price", 233 ...
    ), ...
    struct( ...
        "code", "potato_chips", ...
        "price", 5 ...
    ), ...
    struct( ...
        "code", "towel", ...
        "price", 499 ...
    ) ...
};
sprint("products: ", prettyjsonstringify({products}));

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv1""");

producttofind = "pasta";
sprint("product to find: ", {producttofind});

productfoundindex = arrayfindindexv1(@(product, varargin) strcmp(product.("code"), producttofind), products);
sprint("product found index: ", {productfoundindex});
% product found index: 1

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv2""");

productfoundindex = arrayfindindexv2(@(product, varargin) strcmp(product.("code"), producttofind), products);
sprint("product found index: ", {productfoundindex});
% product found index: 1

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv3""");

productfoundindex = arrayfindindexv3(@(product, varargin) strcmp(product.("code"), producttofind), products);
sprint("product found index: ", {productfoundindex});
% product found index: 1

sprint("% using JavaScript-like Array.findIndex() function ""arrayfindindexv4""");

productfoundindex = arrayfindindexv4(@(product, varargin) strcmp(product.("code"), producttofind), products);
sprint("product found index: ", {productfoundindex});
% product found index: 1

sprint("% using Matlab Array.findIndex() built-in function ""find"" and ""cellfun""");

productfoundindex = find(cellfun(@(product) strcmp(product.("code"), producttofind), products), 1);
sprint("product found index: ", {productfoundindex});
% product found index: 1

File: codes/python/array_find_index.py
import json
from numbers import Number


def array_reduce(callback_function, an_array, initial_value):
    '''JavaScript-like Array.reduce() function'''
    result = initial_value
    for array_item_index, array_item in enumerate(an_array):
        result = callback_function(result, array_item, array_item_index, an_array)
    return result


optional_chaining = lambda anything, *object_properties_array: anything if (((isinstance(anything, dict) == False) and (isinstance(anything, list) == False)) or (len(object_properties_array) == 0)) else array_reduce(lambda current_result, current_item, *_: anything.get(str(current_item)) if ((current_result is None) and (isinstance(anything, dict) == True) and (isinstance(current_item, str) == True)) else anything[int(current_item)] if ((current_result is None) and (isinstance(anything, list) == True) and (isinstance(current_item, Number) == True) and (int(current_item) >= 0) and (len(anything) > int(current_item))) else current_result.get(str(current_item)) if ((isinstance(current_result, dict) == True) and (isinstance(current_item, str) == True)) else current_result[int(current_item)] if ((isinstance(current_result, list) == True) and (isinstance(current_item, Number) == True) and (int(current_item) >= 0) and (len(current_result) > int(current_item))) else None, object_properties_array, None)  # '''JavaScript-like Optional Chaining Operator (?.) function optional_chaining_v3'''

nullish_coalescing = lambda anything, default_value: default_value if (anything is None) else anything  # '''JavaScript-like Nullish Coalescing Operator (??) function nullish_coalescing_v2'''

json_stringify = lambda anything, **optional_argument: (json.dumps(anything, indent=4)) if (nullish_coalescing(optional_chaining(optional_argument, "pretty"), False) == True) else (json.dumps(anything).replace("{", "{ ").replace("}", " }"))

# There's no JavaScript-like Array.findIndex() in Python.
# But, we can create our own function to mimic it in Python.


def array_find_index_v1(callback_function, an_array):
    '''JavaScript-like Array.findIndex() function array_find_index_v1'''
    data_found_index = -1
    for array_item_index, array_item in enumerate(an_array):
        is_condition_match = callback_function(array_item, array_item_index, an_array)
        if (is_condition_match == True):
            data_found_index = array_item_index
            break
    return data_found_index


def array_find_index_v2(callback_function, an_array):
    '''JavaScript-like Array.findIndex() function array_find_index_v2'''
    data_found_index = -1
    for array_item_index, array_item in enumerate(an_array):
        if (callback_function(array_item, array_item_index, an_array) == True):
            data_found_index = array_item_index
            break
    return data_found_index


def array_find_index_v3(callback_function, an_array):
    '''JavaScript-like Array.findIndex() function array_find_index_v3'''
    for array_item_index, array_item in enumerate(an_array):
        is_condition_match = callback_function(array_item, array_item_index, an_array)
        if (is_condition_match == True):
            return array_item_index
    return -1


def array_find_index_v4(callback_function, an_array):
    '''JavaScript-like Array.findIndex() function array_find_index_v4'''
    for array_item_index, array_item in enumerate(an_array):
        if (callback_function(array_item, array_item_index, an_array) == True):
            return array_item_index
    return -1


print("\n# JavaScript-like Array.findIndex() in Python list")

numbers = [12, 34, 27, 23, 65, 93, 36, 87, 4, 254]
print(f'numbers: {json_stringify(numbers)}')

number_to_find = 27
print(f"number to find: {number_to_find}")

print('# using JavaScript-like Array.findIndex() function "array_find_index_v1"')

number_found_index = array_find_index_v1(lambda number, *_: (number == number_to_find), numbers)
print(f"number found index: {number_found_index}")
# number found index: 2

print('# using JavaScript-like Array.findIndex() function "array_find_index_v2"')

number_found_index = array_find_index_v2(lambda number, *_: (number == number_to_find), numbers)
print(f"number found index: {number_found_index}")
# number found index: 2

print('# using JavaScript-like Array.findIndex() function "array_find_index_v3"')

number_found_index = array_find_index_v3(lambda number, *_: (number == number_to_find), numbers)
print(f"number found index: {number_found_index}")
# number found index: 2

print('# using JavaScript-like Array.findIndex() function "array_find_index_v4"')

number_found_index = array_find_index_v4(lambda number, *_: (number == number_to_find), numbers)
print(f"number found index: {number_found_index}")
# number found index: 2

print("\n# JavaScript-like Array.findIndex() in Python list of dictionaries")

products = [
    {
        "code": "pasta",
        "price": 321
    },
    {
        "code": "bubble_gum",
        "price": 233
    },
    {
        "code": "potato_chips",
        "price": 5
    },
    {
        "code": "towel",
        "price": 499
    }
]
print(f"products: {json_stringify(products, pretty=True)}")

product_to_find = 'pasta'
print(f"product to find: {product_to_find}")

print('# using JavaScript-like Array.findIndex() function "array_find_index_v1"')

product_found_index = array_find_index_v1(lambda product, *_: (product['code'] == product_to_find), products)
print(f"product found index: {product_found_index}")
# product found index: 0

print('# using JavaScript-like Array.findIndex() function "array_find_index_v2"')

product_found_index = array_find_index_v2(lambda product, *_: (product['code'] == product_to_find), products)
print(f"product found index: {product_found_index}")
# product found index: 0

print('# using JavaScript-like Array.findIndex() function "array_find_index_v3"')

product_found_index = array_find_index_v3(lambda product, *_: (product['code'] == product_to_find), products)
print(f"product found index: {product_found_index}")
# product found index: 0

print('# using JavaScript-like Array.findIndex() function "array_find_index_v4"')

product_found_index = array_find_index_v4(lambda product, *_: (product['code'] == product_to_find), products)
print(f"product found index: {product_found_index}")
# product found index: 0
